labellines/line_label.py

```diff
@@ -0,0 +1,202 @@
+
+            self.set_path_effects(
+                [
+                    patheffects.Stroke(
+                        linewidth=outline_width, foreground=outline_color
+                    ),
+                    patheffects.Normal(),
+                ]
+            )
+
+        # activate clipping if needed and place on axes
+        if kwargs["clip_on"]:
+            self.set_clip_path(self._ax.patch)
+        self._ax._add_text(self)
+
+    def _update_anchors(self):
+        """
+        This helper method computes the position of the textbox and determines
+        the anchor points needed to adjust the rotation
+        """
+        # Use the mpl-internal float representation (deals with datetime etc)
+        x = self._line.convert_xunits(self._target_x)
+        xdata = self._line.get_xdata(orig=False)
+        ydata = self._line.get_ydata(orig=False)
+
+        mask = np.isfinite(ydata)
+        if mask.sum() == 0:
+            raise ValueError(f"The line {self._line} only contains nan!")
+
+        # Find the first line segment surrounding x
+        for i, (xa, xb) in enumerate(zip(xdata[:-1], xdata[1:])):
+            if min(xa, xb) <= x <= max(xa, xb):
+                ya, yb = ydata[i], ydata[i + 1]
+                break
+        else:
+            raise ValueError("x label location is outside data range!")
+
+        # Interpolate y position of label, (interp needs sorted data)
+        if xa != xb:
+            dx = np.array((xa, xb))
+            dy = np.array((ya, yb))
+            srt = np.argsort(dx)
+            y = np.interp(x, dx[srt], dy[srt])
+        else:  # Vertical case
+            y = (ya + yb) / 2
+
+        # Apply y offset
+        if self._yoffset_logspace:
+            y *= 10 ** self._yoffset
+        else:
+            y += self._yoffset
+
+        if not np.isfinite(y):
+            raise ValueError(
+                f"{self._line} does not have a well defined value"
+                f" at x = {self._target_x}. Consider a different position."
+            )
+
+        self._label_pos = np.array((x, y))
+        self._anchor_a = np.array((xa, ya))
+        self._anchor_b = np.array((xb, yb))
+
+    def _get_rotation(self):
+        # Providing the _rotation property
+        # enables automatic adjustment of the rotation angle
+        # Adapted from https://stackoverflow.com/a/53111799
+        if self._auto_align:
+            # Transform to screen coordinated to make sure the angle is always
+            # correct regardless of axis scaling etc.
+            xa, ya = self._ax.transData.transform(self._anchor_a)
+            xb, yb = self._ax.transData.transform(self._anchor_b)
+            angle = np.rad2deg(np.arctan2(yb - ya, xb - xa))
+
+            # Correct the angle to make sure text is always upright-ish
+            return (angle + 90) % 180 - 90
+        return 0
+
+    def _set_rotation(self, rotation):
+        pass
+
+    _rotation = property(_get_rotation, _set_rotation)
```

Comment:
I am not very comfortable with matplotlib Artists, but this doesn't look very pythonic to me.
Wouldn't it be possible to have instead something like
```python
@property
def _rotation(self):
    ...

@_rotation.setter
def _rotation(self, rotation):
    pass
```